### ISSUE TITLE
Adding app/ to the asset path is a large performance regression

### ISSUE THREAD

**Issue by dhh** [rails/sprockets-rails#234]
Pull request #232 caused a 34% performance hit in our test suite.

**Comment by rafaelfranca** [comment]
Wow! This is huge. I'll take a look

**Comment by josh** [comment]
Woah, yeah, would love to see why. I know an extra search path will cause a little more searching, but not that much.

Your sure its isolated to 21c9c384cd59bf9774f4db7a67d6ec6645e3dfb4? No other changes?

Curious about your integration test suite. Are you doing any parallelization? One thing we make sure we do on the github test suite is warm the asset cache on the prefork hook. Basically just `Rails.application.assets["application.js"]` This ensures we only build it once and theres no chance of separate workers racing to recompute the same data.

**Comment by dhh** [comment]
Just that change. Verified it this morning. Not doing any parallelization
at the moment.

**Comment by josh** [comment]
Can you post your current asset load path set.

```
>> puts Rails.application.assets.paths
```

Wondering if anything else is wrong there.

**Comment by dhh** [comment]
Here's what I get on the version before we expanded the load path:

```
irb(main):001:0> puts Rails.application.assets.paths
/Users/david/37s/apps/bc3/app/assets/fonts
/Users/david/37s/apps/bc3/app/assets/images
/Users/david/37s/apps/bc3/app/assets/javascripts
/Users/david/37s/apps/bc3/app/assets/sounds
/Users/david/37s/apps/bc3/app/assets/stylesheets
/Users/david/37s/apps/bc3/lib/assets/javascripts
/Users/david/37s/apps/bc3/lib/assets/stylesheets
/Users/david/37s/apps/bc3/vendor/assets/javascripts
/Users/david/37s/apps/bc3/vendor/assets/stylesheets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/action_cable-bdbbe18f3cc5/lib/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets/images
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets/stylesheets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/queenbee-plugin-9bad7be658ca/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/file_icons-92381e3ae5ba/app/assets/images
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/file_icons-92381e3ae5ba/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/trix-5116b7a4a10a/vendor/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/infinite_page-33e92758e134/lib/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/gems/local_time-1.0.2/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/turbolinks-e07148b3d46f/lib/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/jquery-rails-272abdd319bb/vendor/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/coffee-rails-c05fb725613d/lib/assets/javascripts
```

And here's what I get after:

```
irb(main):001:0> puts Rails.application.assets.paths
/Users/david/37s/apps/bc3/app/assets/fonts
/Users/david/37s/apps/bc3/app/assets/images
/Users/david/37s/apps/bc3/app/assets/javascripts
/Users/david/37s/apps/bc3/app/assets/sounds
/Users/david/37s/apps/bc3/app/assets/stylesheets
/Users/david/37s/apps/bc3/app/assets
/Users/david/37s/apps/bc3/lib/assets/javascripts
/Users/david/37s/apps/bc3/lib/assets/stylesheets
/Users/david/37s/apps/bc3/vendor/assets/javascripts
/Users/david/37s/apps/bc3/vendor/assets/stylesheets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/action_cable-bdbbe18f3cc5/lib/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets/images
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets/stylesheets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/signal_id-278f61155069/app/assets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/queenbee-plugin-9bad7be658ca/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/queenbee-plugin-9bad7be658ca/app/assets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/file_icons-92381e3ae5ba/app/assets/images
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/file_icons-92381e3ae5ba/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/file_icons-92381e3ae5ba/app/assets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/trix-5116b7a4a10a/vendor/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/infinite_page-33e92758e134/lib/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/gems/local_time-1.0.2/app/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/gems/local_time-1.0.2/app/assets
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/turbolinks-e07148b3d46f/lib/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/jquery-rails-272abdd319bb/vendor/assets/javascripts
/Users/david/.rbenv/versions/2.2.1/lib/ruby/gems/2.2.0/bundler/gems/coffee-rails-c05fb725613d/lib/assets/javascripts
```

**Comment by josh** [comment]
My hunch is most the time wasted is problem this "asset was precompiled" check.

I created a branch that just disables all that code. https://github.com/rails/sprockets-rails/compare/dhh-test

Part of #191 is moving away from the `precompile` regexp/proc support which is hella slow. If we just had a single manifest asset to compare against, the logic becomes `manifest_asset.dependencies.include?(whatever_asset)` which is all cached. Having to iterate over every file in the load path and test `LOOSE_APP_ASSETS` https://github.com/rails/sprockets-rails/blob/master/lib/sprockets/railtie.rb#L45 is just slow.

**Comment by josh** [comment]
If the "precompiled check" can be verified as the problem, I have some ideas for interim patches until we can deprecate the old ways.

1. Cache the result of filter calls. https://github.com/rails/sprockets-rails/blob/master/lib/sprockets/rails/helper.rb#L231 If `app/assets` in the in path, its probably hitting all its subdirectories twice.
2. Port "simple logical path" check up to this level as well. https://github.com/rails/sprockets/pull/12/files if you don't use any Procs or Regexps, we can take a fast path. Unfortunately the default `config.assets.precompile` still uses a Proc, so you can only benefit by clearing that out and using simple paths.

**Comment by dhh** [comment]
Sorry to have let this linger. Let me know what you'd like us to test, and
we're happy to help out. @eileencodes have been helping with a variety of
Rails performance testing with Basecamp, and it'd be great to get her
involved with this, if you still have some ideas you'd like to try.

**Comment by eileencodes** [comment]
Fixed by #265
